### PR TITLE
Skip ArrangeBys in constant folding

### PR DESCRIPTION
**Pull request by ggevay**
The current constant folding removes an `ArrangeBy` when its input is a constant. This is dangerous, because it might happen that a Delta join input is a constant, in which case we lose the `ArrangeBy` that was put there by `JoinImplementation`, which will [frighten](https://github.com/MaterializeInc/materialize/blob/df343265f17c35654fbec59270e76e7c6a45b3d5/src/compute-client/src/plan/mod.rs#L1177) the lowering code.

This PR modifies constant folding to
- not do anything when matching an `ArrangeBy`;
- when looking for constant inputs of other things, look behind `ArrangeBy`s.

Note that this doesn't mean that if we have an `ArrangeBy` somewhere, then constant folding would be completely blocked, or that the `ArrangeBy` would forever remain there. It can still happen that constant folding will remove an `ArrangeBy` when the `ArrangeBy` is the constant input to something else. (In this case, it snatches the constant from under the `ArrangeBy`.) The rationale is that the `ArrangeBy` would have been important only to the thing that was directly on top of it.

Additionally, I changed some other matches on `MirRelationExpr::Constant` to look behind `ArrangeBy`s. (For the ones in `peek.rs` and `sequencer.rs`: I couldn't build a test to trigger these situations when there is an `ArrangeBy` on top of a constant at the top of the query, but I wouldn't be sure it's impossible. Maybe there is a binary join one of whose inputs `is_constant_singleton`, and then `MirRelationExpr::product` is somehow rebuilding that join, and it discards the constant input, and the other input remains there with an `ArrangeBy` on top.)

### Motivation

  * This PR fixes a recognized bug. https://github.com/MaterializeInc/materialize/issues/16195

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - [x] Added the test from Philip's issue.
  - [x] There are some other plan changes. All of these are arranged join inputs, where the `ArrangeBy` was disappearing. (These didn't trigger the panic before, because all of these are Differential joins, whose rendering code is not sensitive to absent `ArrangeBy`s.)
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Fixed an optimizer crash for certain complicated joins.

### CODE REVIEW

**Comment by aalexandrov** [comment]
> It can still happen that constant folding will remove an `ArrangeBy` when the `ArrangeBy` is the constant input to something else.

I'm not sure how this will happen now.

**Comment by ggevay** [comment]
> The only thing I'm slightly worried about are situations where we run join_impls to install some ArrangeBy operators and then eliminate the join altogether.

This should be quite rare: Constant folding itself will not leave the `ArrangeBy`s that were on the inputs when folding a join. (It performs the join on the constants that are under the `ArrangeBy`s, and then leaves a `Constant` node.) Maybe  some other transformation somehow leaves an ArrangeBy that is not necessary, but I couldn't actually construct a test where that happens. And even if some superfluous `ArrangeBy` somehow remains, it shouldn't be a big problem:
- when matching constant expressions, we'll look behind them, so the matches won't fail because of the `ArrangeBy`;
- the lowering should in some cases remove the extra `ArrangeBy`;
- if the lowering doesn't remove it, then we'll pay the price of constructing some extra arrangement.

**Comment by ggevay** [comment]
>> It can still happen that constant folding will remove an ArrangeBy when the ArrangeBy is the constant input to something else.

> I'm not sure how this will happen now.

For example, if there is a Join with `ArrangeBy`s on its inputs, but with all constants under the `ArrangeBy`s, then constant folding the join will remove the `ArrangeBy`s, i.e., the result will just be a `Const` node.

File: src/transform/src/fold_constants.rs [hunk 1]
```diff
@@ -397,22 +389,16 @@ impl FoldConstants {
                     *relation = MirRelationExpr::union_many(new_inputs, relation_type.clone());
                 }
             }
-            MirRelationExpr::ArrangeBy { input, .. } => {
-                if let MirRelationExpr::Constant { .. } = &**input {
-                    *relation = input.take_dangerous();
-                }
+            MirRelationExpr::ArrangeBy { .. } => {
+                // Don't fold ArrangeBys, because that could result in unarranged Delta join inputs.
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/transform/src/fold_constants.rs:393]: Done.

**Pull request merged**